experimental/parser/lex_state.go

```diff
@@ -0,0 +1,165 @@
+// limitations under the License.
+
+package parser
+
+import (
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/token"
+)
+
+// lexer is a Protobuf lexer.
+type lexer struct {
+	token.Context
+	*token.Stream // Embedded so we don't have to call Stream() everywhere.
+	*report.Report
+
+	cursor, count int
+
+	braces []token.ID
+}
+
+func (l *lexer) Push(length int, kind token.Kind) token.Token {
+	l.count++
+	return l.Stream.Push(length, kind)
+}
+
+func (l *lexer) Cursor() int {
+	return l.cursor
+}
+
+// Done returns whether or not we're done lexing runes.
+func (l *lexer) Done() bool {
+	return l.Rest() == ""
+}
+
+// Rest returns unlexed text.
+func (l *lexer) Rest() string {
+	return l.Text()[l.cursor:]
+}
+
+// Peek peeks the next character.
+//
+// Returns -1 if l.Done().
+func (l *lexer) Peek() rune {
+	return decodeRune(l.Rest())
+}
+
+// Pop consumes the next character.
+//
+// Returns -1 if l.Done().
+func (l *lexer) Pop() rune {
+	r := l.Peek()
+	if r != -1 {
+		l.cursor += utf8.RuneLen(r)
+		return r
+	}
+	return -1
+}
+
+// TakeWhile consumes the characters while they match the given function.
+// Returns consumed characters.
+func (l *lexer) TakeWhile(f func(rune) bool) string {
+	start := l.cursor
+	for !l.Done() {
+		r := l.Peek()
+		if r == -1 || !f(r) {
+			break
+		}
+		_ = l.Pop()
+	}
+	return l.Text()[start:l.cursor]
+}
+
+// SeekInclusive seek until the given needle is found; returns the prefix inclusive that
+// needle, and updates the cursor to point after it.
+func (l *lexer) SeekInclusive(needle string) (string, bool) {
+	if idx := strings.Index(l.Rest(), needle); idx != -1 {
+		prefix := l.Rest()[:idx+len(needle)]
+		l.cursor += idx + len(needle)
+		return prefix, true
+	}
+	return "", false
+}
+
+// SeekEOF seeks the cursor to the end of the file and returns the remaining text.
+func (l *lexer) SeekEOF() string {
+	rest := l.Rest()
+	l.cursor += len(rest)
+	return rest
+}
+
+func (l *lexer) Span(start, end int) report.Span {
+	return report.Span{
+		IndexedFile: l.IndexedFile,
+		Start:       start,
+		End:         end,
+	}
+}
+
+func (l *lexer) SpanFrom(start int) report.Span {
+	return l.Span(start, l.cursor)
+}
+
+// HandleICE is a defer-able method that will handle an "internal compiler error"
+// during lexing. This annotates a panic with the lexer state for ease of
+// debugging.
+func (l *lexer) HandleICE() {
+	if panicked := recover(); panicked != nil {
+		panic(fmt.Sprintf("protocompile/parse: panic while lexing {cursor: %d, count: %d}: %s", l.cursor, l.count, panicked))
+	}
+}
+
+// mustProgress returns a progress checker for this lexer.
+func (l *lexer) mustProgress() mustProgress {
+	return mustProgress{l, -1}
+}
+
+// mustProgress is a helper for ensuring that the lexer makes progress
+// in each loop iteration. This is intended for turning infinite loops into
+// panics.
```

Comment:
This is intended to convert infinite loops into crashes so that they can be reported as bugs. It's code that should never be exercised, but exists in case there is a corner case we didn't find.

It's also very useful for development.